lib/common/xpath.c

```diff
@@ -266,37 +266,65 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
     return result;
 }
 
-int
-pcmk__element_xpath(const xmlNode *xml, char *buffer, int offset,
-                    size_t buffer_size)
+/*!
+ * \internal
+ * \brief Get an XPath string that matches an XML element as closely as possible
+ *
+ * \param[in] xml  The XML element for which to build an XPath string
+ *
+ * \return A \p GString that matches \p xml, or \p NULL if \p xml is \p NULL.
+ *
+ * \note The caller is responsible for freeing the string using
+ *       \p g_string_free().
+ */
+GString *
+pcmk__element_xpath(const xmlNode *xml)
 {
-    const char *id = ID(xml);
+    const char *id = NULL;
+    GString *xpath = NULL;
 
-    if ((offset == 0) && (xml->parent != NULL)) {
-        offset = pcmk__element_xpath(xml->parent, buffer, offset, buffer_size);
+    // The document node has no parent and no name
+    if (xml->parent == NULL) {
+        return NULL;
```

Comment:
This needs a little tune-up... if we pass in the document node, we get `NULL` back.

Reply:
GitHub doesn't show my comment as outdated but it's referring to a previous version of this function. I think the function as written now should work fine for any node.

If we want, I can definitely still restrict it to non-document in the doxygen and modify the code accordingly. It would consolidate two `g_string_*()` calls into one when (and only when) there's an ID.

Reply:
Not worth changing